### PR TITLE
Fix checkPlatform takes an insane amount of time to complete when used in a loop

### PR DESCRIPTION
**Pull request by Tofandel**
Yes this is a bit dirty because the report is now generated once at the beginning of a command and cached in memory (but normally treeshaking will only include it only if needed)

But somehow when the call is done in the beginning of the process, this call is very fast, but when it's called after having run a lot of code (as is currently the case, a single call takes exponentially more time to complete, it takes 26s on my system when called at the end of npm upgrade) but only a few milliseconds when called at the beginning

This fixes npm hanging https://github.com/npm/cli/issues/4028, https://github.com/npm/cli/issues/7814

### CODE REVIEW

**Comment by IngFu** [comment]
> Another option is to copy the logic from yarn which also fixed this issue by looking into the `/usr/bin/ldd` file and fallback to getReport after
> 
> https://github.com/yarnpkg/berry/blob/f59bbf9f3828865c14b06a3e5cc3ae284a0db78d/packages/yarnpkg-core/sources/nodeUtils.ts#L28-L67

Is getReport() called once during npm install or multiple times?

**Comment by Tofandel** [comment]
Currently multiple times because it's called at every checkPlatform, which is done in a loop once per installed package during npm install

**Comment by Tofandel** [comment]
@reggi Any chance you can review and merge this quickly as this issue is serious (can completely hog CI servers) and prevents `npm install` or `npm ugrade` from working and it's really affecting a lot of people
https://github.com/nodejs/docker-node/issues/1946
https://github.com/npm/cli/issues/7891
https://github.com/npm/cli/issues/7900
https://github.com/npm/cli/issues/7868
https://github.com/npm/cli/issues/4028
https://github.com/npm/cli/issues/7814

**Comment by wraithgar** [comment]
I don't think the approach of assigning our own attributes to `process.report` is wise.  Ideally if we want to cache these values, we'd move that to a separate file that we can then mock in the tests to bypass the cache.

**Comment by Tofandel** [comment]
@wraithgar I added the ldd file check with tests and use a mocked file to bypass the cache and avoid assigning to process.report

File: lib/current-env.js [hunk 1]
```diff
@@ -13,20 +14,42 @@ function cpu () {
...
-  let family
-  const originalExclude = process.report.excludeNetwork
-  process.report.excludeNetwork = true
-  const report = process.report.getReport()
-  process.report.excludeNetwork = originalExclude
-  if (report.header?.glibcVersionRuntime) {
-    family = 'glibc'
-  } else if (Array.isArray(report.sharedObjects) && report.sharedObjects.some(isMusl)) {
-    family = 'musl'
+  if (family === undefined) {
```
Comment (wraithgar) [inline lib/current-env.js:38]: We'll need a simple test added that *doesn't* cache bypass so we can test the case where family is not undefined.

**Comment by wraithgar** [comment]
We're VERY close imo.  Just gotta get the coverage fixed and we can ship this.  I'm holding off on the [next npm release](https://github.com/npm/cli/pull/7923) till this lands because I think it should be pretty soon.

File: lib/current-env.js [hunk 2]
```diff
@@ -13,20 +14,42 @@ function cpu () {
...
+    family = getFamilyFromFilesystem()
+    if (family === undefined) {
```
Comment (styfle) [inline lib/current-env.js:40]: Perhaps the contents of this `if` block should be moved to a new function?

Something like `getFamilyFromReport()`

**Comment by wraithgar** [comment]
The testing setup is a little hard to follow with the extra callbacks but I don't think it's a blocker. We can refactor that later.

pinging @H4ad in case they had any input before we land this (hopefully today)

**Review by wraithgar — APPROVED** [review]

**Review by H4ad — COMMENTED** [review]
I can't approve but LGTM!

**Comment by wraithgar** [comment]
Huge thanks to @H4ad @styfle and @Tofandel.  There was a lot of coordination done to tie all the similar issues together, and of course to get this fix out the door.
